raiden/network/proxies/token_network_registry.py

```diff
@@ -113,107 +115,332 @@ def add_token(
+            )
+
+        if channel_participant_deposit_limit > token_network_deposit_limit:
+            raise InvalidChannelParticipantDepositLimit(
+                f"Channel participant deposit limit of "
+                f"{channel_participant_deposit_limit} is invalid"
+            )
+
+        token_proxy = self.blockchain_service.token(token_address)
         try:
+            token_supply = token_proxy.total_supply(block_identifier=block_identifier)
             already_registered = self.get_token_network(
                 token_address=token_address, block_identifier=block_identifier
             )
+            deprecation_executor = self.get_deprecation_executor(block_identifier=block_identifier)
+            settlement_timeout_min = self.settlement_timeout_min(block_identifier=block_identifier)
+            settlement_timeout_max = self.settlement_timeout_max(block_identifier=block_identifier)
+            chain_id = self.get_chain_id(block_identifier=block_identifier)
+            secret_registry_address = self.get_secret_registry_address(
+                block_identifier=block_identifier
+            )
+            max_token_networks = self.get_max_token_networks(block_identifier=block_identifier)
+            token_networks_created = self.get_token_network_created(
+                block_identifier=block_identifier
+            )
         except ValueError:
             # If `block_identifier` has been pruned the checks cannot be performed
             pass
         except BadFunctionCallOutput:
             raise_on_call_returned_empty(block_identifier)
         else:
+            if token_networks_created + 1 >= max_token_networks:
+                raise BrokenPreconditionError(
+                    f"Number of token networks will exceed the max of {max_token_networks}"
+                )
+
+            if token_supply == "":
+                raise InvalidToken(
+                    "Given token address does not follow the "
+                    "ERC20 standard (missing `totalSupply()`)"
+                )
             if already_registered:
                 raise BrokenPreconditionError(
                     "The token is already registered in the TokenNetworkRegistry."
                 )
 
-        return self._add_token(
-            token_address=token_address,
-            channel_participant_deposit_limit=channel_participant_deposit_limit,
-            token_network_deposit_limit=token_network_deposit_limit,
-        )
+            if deprecation_executor == NULL_ADDRESS:
+                raise BrokenPreconditionError(
+                    "The deprecation executor property for the TokenNetworkRegistry is invalid."
+                )
+
+            if chain_id == 0:
+                raise BrokenPreconditionError(
+                    "The chain ID property for the TokenNetworkRegistry is invalid."
+                )
+
+            if chain_id != self.blockchain_service.network_id:
```

Comment:
I think we should not check this.  After a fork, there will be a fork with the original `chain_id` and another fork with a different `chain_id`.  The choice is on the user.

Reply:
That's not something we have to handle. Because when the user runs Raiden, it is ran on a specific chain ID. If there was a fork at some point and the user choose to use that new chain, Raiden has to be restarted to use that chain which will have a new ID.

Reply:
1. We deploy the contract on `chain_id == 1` and store `chain_id == 1` in the contract
2. A fork happens into `chain_id`s 1 and 231.
3. A user tries to use the smart contract on chain 231.

In this case, the smart contract still remembers `chain_id == 1` (that's the left hand side).  But `blockchain_service.network_id` points to 231 because now the user is trying to use chain_id 231.  So the check fails here.  If the proxy allows, the transaction would succeed onchain because on-chain we don't check that the `chain_id` is a particular value.

Reply:
I already know that the user runs Raiden on a specific chain ID, but I see the problem ^.

Reply:
Isn't it also a problem that we store the chain id in the contract though? If there is a fork and both sides survive (let's say 1 and 231 per your example), the contracts would need to be redeployed in 231 chain.

If they are not then the on-chain `chain_id` would be wrong and as such all functions using it such as `recoverAddressFromBalanceProof()` would not work.

Reply:
Currently, the onchain `chain_id` would be wrong.  That's why no strict checks should be performed with it.